#Import Lib

In [1]:
from pandas import read_csv # membaca csv
from sklearn.naive_bayes import GaussianNB # untuk membuat model Naive Bayes
from sklearn.tree import DecisionTreeClassifier # untuk membuat model decision tree
from sklearn.model_selection import train_test_split # split data latih dengan data test
from sklearn import preprocessing
from itertools import cycle

#Data Preparation


Load data

In [2]:
data = read_csv('sample_data/IRIS.csv')
data.head()

sepal_length  sepal_width  petal_length  petal_width      species
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

Label encoding

In [3]:
# label encoding digunakan ketika attribute target prediksi bersifat kategorial
label_encoder = preprocessing.LabelEncoder()
 
# Encoding label pada kolom 'species'.
data['species']= label_encoder.fit_transform(data['species'])
# menjadi: iris-setosa = 0, iris-versicolor = 1, iris-virginica = 2
data.sample(8)

sepal_length  sepal_width  petal_length  petal_width  species
66            5.6          3.0           4.5          1.5        1
143           6.8          3.2           5.9          2.3        2
48            5.3          3.7           1.5          0.2        0
108           6.7          2.5           5.8          1.8        2
90            5.5          2.6           4.4          1.2        1
22            4.6          3.6           1.0          0.2        0
67            5.8          2.7           4.1          1.0        1
0             5.1          3.5           1.4          0.2        0

#Main

Fungsi untuk melatih dan menguji model menggunakan data

In [4]:
def main(data, model):
    #memisahkan target dan pendukung (x dan y)
    cykolom = cycle(data.columns.to_list())
    x = data[list(next(cykolom) for _ in range(4))] # pendukung
    y = data.species # Target 
    model = eval(x, y, model)

    print("user input:")
    input_data = [5.9, 3.1, 5.1, 1.9]
    for i in range(4): print(f'{data.columns.to_list()[i]}\t: {input_data[i]}')
    print(f'species\t\t: {prediksi(input_data, model)}')

Fungsi untuk menghitung akurasi model

In [5]:
sama = lambda x, y, i: x[i] == y[i]
def akurasi(asli, prediksi):
	benar = 0
	for i in range(len(asli)):
		if sama(asli, prediksi, i): benar += 1
	yield benar / float(len(asli)) * 100

Mengembalikan nilai yang diprediksi

In [6]:
def prediksi(data, tree):
    if tree.predict([data]) == 0: return 'Iris-setosa'
    elif tree.predict([data]) == 1: return 'Iris-versicolor'
    else : return 'Iris-virginica' 

In [7]:
def eval(x, y, model):
    # 70% data untuk pelatihan, 30% untuk tes, dan random state = 1
    x_latih, x_tes, y_latih, y_tes =  train_test_split(x, y, test_size=0.3, random_state=1) 
    
    # melatih model
    model = model.fit(x_latih,y_latih)
    
    #prediksi untuk melihat akurasi
    y_pred = model.predict(x_tes)

    # menampilkan akurasi
    acc = akurasi(list(y_tes), list(y_pred))
    print("akurasi: %.2f%%\n" % (next(acc)))
    return model

#Decision Tree

In [8]:
main(data, DecisionTreeClassifier(criterion="gini", max_depth=3))

akurasi: 95.56%

user input:
sepal_length	: 5.9
sepal_width	: 3.1
petal_length	: 5.1
petal_width	: 1.9
species		: Iris-virginica


untuk model Decision Tree didapatkan akurasi 95.56% dengan max depth 3. kelebihan dari DT adalah dapat menguji data numerik dan kategorial, namun kelemahan DT saat menjadi tree yang kompleks, variasi data yang kecil dapat menyebabkan hasil data yang berbeda 

#Naïve Bayes Classifier 

In [9]:
main(data, GaussianNB())

akurasi: 93.33%

user input:
sepal_length	: 5.9
sepal_width	: 3.1
petal_length	: 5.1
petal_width	: 1.9
species		: Iris-virginica


pada model Naive bayes didapatkan akurasi 93.33%. Model NB memiliki kelebihan dalam pelatihan data yang memerlukan sedikit data latih serta unggul dalam hal kecepatan. NB hanya berdasarkan asumsi, independensi bergantung pada kelasnya, akurasinya berkurang

#Kesimpulan

Pada perbandingan dua model di atas, dapat dipilih sesuai kelebihan dan kekurangan masing-masing. Jika ingin mementingkan akurasi dan visualisasi Kita dapat menggunakan Decision Tree. Kemudian jika lebih memerlukan kecepatan atau data yang dimiliki terbatas, menggunakan Naive Bayes adalah solusinya. 

source: 
- https://discuss.analyticsvidhya.com/t/decision-tree-vs-naive-bayes-classifier/2520
- https://analyticsindiamag.com/7-types-classification-algorithms/